custom_components/rfplayer/rflib/rfpparser.py

```diff
@@ -90,65 +28,47 @@ def valid_packet(packet: str) -> bool:
 def decode_packet(packet: str) -> list:
     """Decode packet."""
     packets_found = []
-    data = cast(PacketType, {"node": PacketHeader.GATEWAY.name})
+    data = {}
 
-    # Welcome messages directly send
+    # Response to any command
     if packet.startswith("ZIA--"):
-        data["message"] = packet.replace("ZIA--", "")
+        data["response"] = packet.replace("ZIA--", "")
         return [data]
 
     # Protocols
     message = json.loads(packet.replace("ZIA33", ""))["frame"]
     data["protocol"] = message["header"]["protocolMeaning"]
 
-    if data["protocol"] in ["BLYSS", "CHACON", "JAMMING"]:
-        data["id"] = message["infos"]["id"]
-        data["command"] = message["infos"]["subType"]
+    # FIXME improve typing
+    if data["protocol"] in ["JAMMING"]:
+        data["device_id"] = message["infos"]["id"]
+        data["command"] = "jamming_detection"
+        data["state"] = message["infos"]["subType"]
+        packets_found.append(data)
+    if data["protocol"] in ["BLYSS", "CHACON"]:
+        data["device_id"] = message["infos"]["id"]
+        data["command"] = "cmd"
         data["state"] = message["infos"]["subTypeMeaning"]
         packets_found.append(data)
     elif data["protocol"] in ["X2D"]:
-        data["id"] = message["infos"]["id"]
-        if message["infos"]["subTypeMeaning"] == "Detector/Sensor":
-            value = VALUE_TRANSLATION["detector"](message["infos"]["qualifier"])
-            data["command"] = value
-            data["state"] = value
-        elif message["infos"]["subTypeMeaning"] == "STARBOX F03":
```

Comment:
we don't lost some protocol support ?

Reply:
don't know about all X2D variants. but seems like it can be either a detector or thermostat. thermostats are using a complex command which cannot be mapped to a simple on / off switch like lights. So I decided to report the current thermostat state as a sensor and use send_command with right format to adjust the operating mode / heating speed.
A better option we'd be to map X2D thermostat to a climate entity cause it supports both on/off + presets. But it's more work.

visionic was not stored in the list so it was actually already unsupported.

anyway I didn't plan to make all protocols work cause there was no test with payloads from the different protocols and since I don't have real devices for all the rf protocols supported by rfplayer I can't test with real devices.

Reply:
I'm afraid this change breaks X2D support:

sensor / detector protocol messages do not contain "functionMeaning" or "state" fields:

message:{'header': {'frameType': '0', 'cluster': '0', 'dataFlag': '1', 'rfLevel': '-91', 'floorNoise': '-105', 'rfQuality': '3', 'protocol': '8', 'protocolMeaning': 'X2D', 'infoType': '11', 'frequency': '868350'}, 'infos': {'subType': '0', 'subTypeMeaning': 'Detector/Sensor', 'id': '2916083969', 'qualifier': '2', 'qualifierMeaning': {'flags': ['Alarm']}, 'd0': '55804', 'd1': '0', 'd2': '0', 'd3': '0'}}

and information needs to be taken from "qualifier" field (then left as numerical value or mapped as it was before ... should be same as for other protocols I guess...) and "flags" (note: flags is empty - [] - when semsor reports 'normal' state (ie: opening detector closed).

Starbox (the 'thermostat') messages are different:

message:{'header': {'frameType': '0', 'cluster': '0', 'dataFlag': '1', 'rfLevel': '-54', 'floorNoise': '-105', 'rfQuality': '10', 'protocol': '8', 'protocolMeaning': 'X2D', 'infoType': '10', 'frequency': '868350'}, 'infos': {'subType': '6', 'subTypeMeaning': 'STARBOX F03', 'id': '604702208', 'area': '0', 'qualifier': '32', 'qualifierMeaning': {'flags': []}, 'function': '2', 'functionMeaning': 'OPERATING MODE', 'state': '7', 'stateMeaning': 'AUTO', 'd0': '1794', 'd1': '2085', 'd2': '63742', 'd3': '0'}}


Type of message is identified by "subType" / "subTypeMeaning" - and there are more than "Detector/Sensor" and "STARBOX F03" types (ie: shutters/covers ... etc) ... but I only have these two ... so no idea what others are reporting ... 











T

Reply:
ok fair enough I did that quickly. will try to restore X2D support for other device types.
but AFAIU your sensor / detector was mapped as a switch ?

Reply:
qualifier is a bitmap and the previous code with DTC_STATUS_LOOKUP assumes it's some sort of enum. don't think it worked properly. to make it work, it should be broken down into multiple entities one for each bit of the map. but I'll keep it simple with single sensor entity with the qualifier value

Reply:
@jaroslawp can you have a look at the new commit and let me know WDYT

Reply:
Yes, that should work well for my X2Ds - thank you!